src/goto-programs/destructor_tree.cpp

```diff
@@ -0,0 +1,108 @@
+#include "destructor_tree.h"
+
+/// Adds a destructor to the current stack, either attaching itself to the
+/// closest previous node, or an appropriate rewind target.
+void destructor_treet::add(const codet &destructor)
+{
+  // If we have a stored rewind target, attach node to that then wipe rewind.
+  auto previous_node = get_current_node();
+  auto new_node = destruction_graph.add_node(destructor);
+  if(rewind_target)
+  {
+    auto rewind_target_index = *rewind_target;
+    rewind_target = {};
+    if(!destruction_graph.has_edge(rewind_target_index, new_node))
+      destruction_graph.add_edge(rewind_target_index, new_node);
+  }
+  else
+  {
+    destruction_graph.add_edge(previous_node, new_node);
+  }
+}
+
+/// Fetches the destructor value for the passed-in node identity.
```

Comment:
Suggest `identity` (which suggests an identity function perhaps?) for index everywhere, since that's the term used in `grapht` (node index)